### PR TITLE
[dynamo][activation checkpointing] Trace through ActivationWrapper

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108599

Fixes https://github.com/pytorch/pytorch/issues/108269

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: test/distributed/test_dynamo_distributed.py [hunk 1]
```diff
@@ -302,6 +302,46 @@ def test_hf_bert_ddp_aot_eager(self):
...
+                checkpoint_impl=CheckpointImpl.NO_REENTRANT,
+            )
+            check_fn = lambda submodule: isinstance(submodule, torch.nn.Linear)  # noqa: E731
+            apply_activation_checkpointing(model, checkpoint_wrapper_fn=non_reentrant_wrapper, check_fn=check_fn)
+
+            model = DDP(model)
+            x = torch.randn(10, 64).cuda()
+            correct_outputs = model(x)
+
+            opt_model = torch.compile(model)
```
Comment (rohan-varma) [inline test/distributed/test_dynamo_distributed.py:341]: do we also wanna test applying AC after DDP wrap?

**Review by rohan-varma — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting your change, but test_ddp_activation_checkpointing is failing distributed ROCm test in trunk' -c nosignal

Here is an example failure https://hud.pytorch.org/pytorch/pytorch/commit/43527d41a255322003fd806eaefce56f0de7bfb0, where the test time out:

```
2023-09-07T02:23:27.3516428Z _______________ TestMultiProc.test_ddp_activation_checkpointing ________________
2023-09-07T02:23:27.3516786Z Traceback (most recent call last):
2023-09-07T02:23:27.3517351Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 542, in wrapper
2023-09-07T02:23:27.3517770Z     self._join_processes(fn)
2023-09-07T02:23:27.3518308Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 761, in _join_processes
2023-09-07T02:23:27.3518761Z     self._check_return_codes(elapsed_time)
2023-09-07T02:23:27.3519343Z   File "/opt/conda/envs/py_3.8/lib/python3.8/site-packages/torch/testing/_internal/common_distributed.py", line 816, in _check_return_codes
2023-09-07T02:23:27.3519769Z     raise RuntimeError(
2023-09-07T02:23:27.3520110Z RuntimeError: Process 0 terminated or timed out after 310.09846544265747 seconds
2023-09-07T02:23:27.3520585Z ----------------------------- Captured stdout call -----------------------------
2023-09-07T02:23:27.3520973Z Timing out after 300 seconds and killing subprocesses.
```

**Pull request closed without merging**